### PR TITLE
chore: use docker bake to generate docker compose file

### PR DESCRIPTION
**Pull request by lemeurherve**
This PR uses docker bake to generate the content of the docker compose files for Windows images from the same docker-bake.hcl used for Linux images.

Fixes:
- #426

The JSON output of `docker buildx bake windows --print` is converted with `yq` into the YAML format expected by docker compose to build the corresponding images. (We can't use bake for building as it's not functional on Windows yet.)

~The docker compose files are generated (here) from a Linux agent as it avoids tedious `yq` queries escaping for PowerShell and installing docker buildx on Windows agents.~

EDIT:
As `docker buildx bake windows --print` can be used on Windows, `make` isn't required for contributors. 
And as `yq` is already required, the only new requirement for contributors is https://github.com/docker/buildx included in Docker Desktop, one of the most common docker installation types on Windows.

<details><summary>Example of resulting build-windows.yaml (note that no arg depends on any variable)</summary>

```yaml
services:
  nanoserver-1809_jdk11:
    image: docker.io/jenkins/ssh-agent:1.2.3-nanoserver-1809-jdk11
    build:
      context: .
      dockerfile: windows/nanoserver/Dockerfile
      args:
        JAVA_HOME: C:/openjdk-11
        JAVA_VERSION: 11.0.23+9
        TOOLS_WINDOWS_VERSION: "1809"
        WINDOWS_VERSION_TAG: "1809"
      tags:
        - docker.io/jenkins/ssh-agent:1.2.3-nanoserver-1809-jdk11
        - docker.io/jenkins/ssh-agent:nanoserver-1809-jdk11
  nanoserver-1809_jdk17:
    image: docker.io/jenkins/ssh-agent:1.2.3-nanoserver-1809-jdk17
    build:
      context: .
      dockerfile: windows/nanoserver/Dockerfile
      args:
        JAVA_HOME: C:/openjdk-17
        JAVA_VERSION: 17.0.11+9
        TOOLS_WINDOWS_VERSION: "1809"
        WINDOWS_VERSION_TAG: "1809"
      tags:
        - docker.io/jenkins/ssh-agent:1.2.3-nanoserver-1809-jdk17
        - docker.io/jenkins/ssh-agent:1.2.3-nanoserver-1809
        - docker.io/jenkins/ssh-agent:nanoserver-1809
        - docker.io/jenkins/ssh-agent:nanoserver-1809-jdk17
  nanoserver-1809_jdk21:
    image: docker.io/jenkins/ssh-agent:1.2.3-nanoserver-1809-jdk21
    build:
      context: .
      dockerfile: windows/nanoserver/Dockerfile
      args:
        JAVA_HOME: C:/openjdk-21
        JAVA_VERSION: 21.0.3+9
        TOOLS_WINDOWS_VERSION: "1809"
        WINDOWS_VERSION_TAG: "1809"
      tags:
        - docker.io/jenkins/ssh-agent:1.2.3-nanoserver-1809-jdk21
        - docker.io/jenkins/ssh-agent:nanoserver-1809-jdk21
  nanoserver-ltsc2019_jdk11:
    image: docker.io/jenkins/ssh-agent:1.2.3-nanoserver-ltsc2019-jdk11
    build:
      context: .
      dockerfile: windows/nanoserver/Dockerfile
      args:
        JAVA_HOME: C:/openjdk-11
        JAVA_VERSION: 11.0.23+9
        TOOLS_WINDOWS_VERSION: "1809"
        WINDOWS_VERSION_TAG: ltsc2019
      tags:
        - docker.io/jenkins/ssh-agent:1.2.3-nanoserver-ltsc2019-jdk11
        - docker.io/jenkins/ssh-agent:nanoserver-ltsc2019-jdk11
  nanoserver-ltsc2019_jdk17:
    image: docker.io/jenkins/ssh-agent:1.2.3-nanoserver-ltsc2019-jdk17
    build:
      context: .
      dockerfile: windows/nanoserver/Dockerfile
      args:
        JAVA_HOME: C:/openjdk-17
        JAVA_VERSION: 17.0.11+9
        TOOLS_WINDOWS_VERSION: "1809"
        WINDOWS_VERSION_TAG: ltsc2019
      tags:
        - docker.io/jenkins/ssh-agent:1.2.3-nanoserver-ltsc2019-jdk17
        - docker.io/jenkins/ssh-agent:1.2.3-nanoserver-ltsc2019
        - docker.io/jenkins/ssh-agent:nanoserver-ltsc2019
        - docker.io/jenkins/ssh-agent:nanoserver-ltsc2019-jdk17
  nanoserver-ltsc2019_jdk21:
    image: docker.io/jenkins/ssh-agent:1.2.3-nanoserver-ltsc2019-jdk21
    build:
      context: .
      dockerfile: windows/nanoserver/Dockerfile
      args:
        JAVA_HOME: C:/openjdk-21
        JAVA_VERSION: 21.0.3+9
        TOOLS_WINDOWS_VERSION: "1809"
        WINDOWS_VERSION_TAG: ltsc2019
      tags:
        - docker.io/jenkins/ssh-agent:1.2.3-nanoserver-ltsc2019-jdk21
        - docker.io/jenkins/ssh-agent:nanoserver-ltsc2019-jdk21
  nanoserver-ltsc2022_jdk11:
    image: docker.io/jenkins/ssh-agent:1.2.3-nanoserver-ltsc2022-jdk11
    build:
      context: .
      dockerfile: windows/nanoserver/Dockerfile
      args:
        JAVA_HOME: C:/openjdk-11
        JAVA_VERSION: 11.0.23+9
        TOOLS_WINDOWS_VERSION: ltsc2022
        WINDOWS_VERSION_TAG: ltsc2022
      tags:
        - docker.io/jenkins/ssh-agent:1.2.3-nanoserver-ltsc2022-jdk11
        - docker.io/jenkins/ssh-agent:nanoserver-ltsc2022-jdk11
  nanoserver-ltsc2022_jdk17:
    image: docker.io/jenkins/ssh-agent:1.2.3-nanoserver-ltsc2022-jdk17
    build:
      context: .
      dockerfile: windows/nanoserver/Dockerfile
      args:
        JAVA_HOME: C:/openjdk-17
        JAVA_VERSION: 17.0.11+9
        TOOLS_WINDOWS_VERSION: ltsc2022
        WINDOWS_VERSION_TAG: ltsc2022
      tags:
        - docker.io/jenkins/ssh-agent:1.2.3-nanoserver-ltsc2022-jdk17
        - docker.io/jenkins/ssh-agent:1.2.3-nanoserver-ltsc2022
        - docker.io/jenkins/ssh-agent:nanoserver-ltsc2022
        - docker.io/jenkins/ssh-agent:nanoserver-ltsc2022-jdk17
  nanoserver-ltsc2022_jdk21:
    image: docker.io/jenkins/ssh-agent:1.2.3-nanoserver-ltsc2022-jdk21
    build:
      context: .
      dockerfile: windows/nanoserver/Dockerfile
      args:
        JAVA_HOME: C:/openjdk-21
        JAVA_VERSION: 21.0.3+9
        TOOLS_WINDOWS_VERSION: ltsc2022
        WINDOWS_VERSION_TAG: ltsc2022
      tags:
        - docker.io/jenkins/ssh-agent:1.2.3-nanoserver-ltsc2022-jdk21
        - docker.io/jenkins/ssh-agent:nanoserver-ltsc2022-jdk21
  windowsservercore-ltsc2019_jdk11:
    image: docker.io/jenkins/ssh-agent:1.2.3-windowsservercore-ltsc2019-jdk11
    build:
      context: .
      dockerfile: windows/windowsservercore/Dockerfile
      args:
        JAVA_HOME: C:/openjdk-11
        JAVA_VERSION: 11.0.23+9
        TOOLS_WINDOWS_VERSION: "1809"
        WINDOWS_VERSION_TAG: ltsc2019
      tags:
        - docker.io/jenkins/ssh-agent:1.2.3-windowsservercore-ltsc2019-jdk11
        - docker.io/jenkins/ssh-agent:windowsservercore-ltsc2019-jdk11
  windowsservercore-ltsc2019_jdk17:
    image: docker.io/jenkins/ssh-agent:1.2.3-windowsservercore-ltsc2019-jdk17
    build:
      context: .
      dockerfile: windows/windowsservercore/Dockerfile
      args:
        JAVA_HOME: C:/openjdk-17
        JAVA_VERSION: 17.0.11+9
        TOOLS_WINDOWS_VERSION: "1809"
        WINDOWS_VERSION_TAG: ltsc2019
      tags:
        - docker.io/jenkins/ssh-agent:1.2.3-windowsservercore-ltsc2019-jdk17
        - docker.io/jenkins/ssh-agent:1.2.3-windowsservercore-ltsc2019
        - docker.io/jenkins/ssh-agent:windowsservercore-ltsc2019
        - docker.io/jenkins/ssh-agent:windowsservercore-ltsc2019-jdk17
  windowsservercore-ltsc2019_jdk21:
    image: docker.io/jenkins/ssh-agent:1.2.3-windowsservercore-ltsc2019-jdk21
    build:
      context: .
      dockerfile: windows/windowsservercore/Dockerfile
      args:
        JAVA_HOME: C:/openjdk-21
        JAVA_VERSION: 21.0.3+9
        TOOLS_WINDOWS_VERSION: "1809"
        WINDOWS_VERSION_TAG: ltsc2019
      tags:
        - docker.io/jenkins/ssh-agent:1.2.3-windowsservercore-ltsc2019-jdk21
        - docker.io/jenkins/ssh-agent:windowsservercore-ltsc2019-jdk21
  windowsservercore-ltsc2022_jdk11:
    image: docker.io/jenkins/ssh-agent:1.2.3-windowsservercore-ltsc2022-jdk11
    build:
      context: .
      dockerfile: windows/windowsservercore/Dockerfile
      args:
        JAVA_HOME: C:/openjdk-11
        JAVA_VERSION: 11.0.23+9
        TOOLS_WINDOWS_VERSION: ltsc2022
        WINDOWS_VERSION_TAG: ltsc2022
      tags:
        - docker.io/jenkins/ssh-agent:1.2.3-windowsservercore-ltsc2022-jdk11
        - docker.io/jenkins/ssh-agent:windowsservercore-ltsc2022-jdk11
  windowsservercore-ltsc2022_jdk17:
    image: docker.io/jenkins/ssh-agent:1.2.3-windowsservercore-ltsc2022-jdk17
    build:
      context: .
      dockerfile: windows/windowsservercore/Dockerfile
      args:
        JAVA_HOME: C:/openjdk-17
        JAVA_VERSION: 17.0.11+9
        TOOLS_WINDOWS_VERSION: ltsc2022
        WINDOWS_VERSION_TAG: ltsc2022
      tags:
        - docker.io/jenkins/ssh-agent:1.2.3-windowsservercore-ltsc2022-jdk17
        - docker.io/jenkins/ssh-agent:1.2.3-windowsservercore-ltsc2022
        - docker.io/jenkins/ssh-agent:windowsservercore-ltsc2022
        - docker.io/jenkins/ssh-agent:windowsservercore-ltsc2022-jdk17
  windowsservercore-ltsc2022_jdk21:
    image: docker.io/jenkins/ssh-agent:1.2.3-windowsservercore-ltsc2022-jdk21
    build:
      context: .
      dockerfile: windows/windowsservercore/Dockerfile
      args:
        JAVA_HOME: C:/openjdk-21
        JAVA_VERSION: 21.0.3+9
        TOOLS_WINDOWS_VERSION: ltsc2022
        WINDOWS_VERSION_TAG: ltsc2022
      tags:
        - docker.io/jenkins/ssh-agent:1.2.3-windowsservercore-ltsc2022-jdk21
        - docker.io/jenkins/ssh-agent:windowsservercore-ltsc2022-jdk21
```

Obtained with:
```console
ON_TAG=true VERSION=1.2.3 docker buildx bake -f docker-bake.hcl windows --print \
| yq -P '.target[] | del(.platforms) | {(. | key): {"image": .tags[0], "build": .}}'
| yq '{"services": .}'
```
Note: this command works for Linux/MacOS, Windows (or more specifically PowerShell) requires backslashes in front of every double quote and setting environment variables on their own.

</details>

Enhancements:
- Regroup all Linux and Windows image definitions in a common docker-bake.hcl file containing every parameters and helper functions
- Remove build-windows.yaml
- Simplify Jenkinsfile
- Simplify build.ps1 by not having to massage parameters to deal with the fact that there are no mcr.microsoft.com/powershell ltsc2019 images, only 1809 ones
- Simplify updatecli manifests
- Prepare the terrain for when docker buildx will be functional to build Windows images

Images content remains untouched hence the "chore" label.

### Testing done

Local tests and CI.

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Review by dduportal — COMMENTED** [review]
Could you update https://github.com/jenkinsci/docker-ssh-agent?tab=readme-ov-file#building-and-testing-on-windows as part of this PR please?

Given 

> The docker compose files are generated from a Linux agent as it avoids tedious yq queries escaping for PowerShell and installing docker buildx on Windows agents.

It looks like you need a lot of additional tools if you are on Windows to be able to build the image (including GNU `make` and `yq`): am I correct? If yes, wether we agree or not, this PR need the doc. to be updated to guide contributors

File: docker-bake.hcl
Comment (dduportal) [inline docker-bake.hcl:94]: Could you explain why are there backslash on these? It's not really clear as I would suggest to write as a [normal HCL list](https://developer.hashicorp.com/terraform/language/expressions/types#list) such as:

```suggestion
  default = ["nanoserver","windowsservercore"]
```

File: docker-bake.hcl
Comment (dduportal) [inline docker-bake.hcl:98]: Same as https://github.com/jenkinsci/docker-ssh-agent/pull/415/files#r1632209159

**Review by dduportal — COMMENTED** [review]
This is a really interesting PR technically speaking, showing a lot of different usages and improvements over the current HCL usage.

- There are a lot of tiny improvement which looks really useful without any doubts, such as factorizations to variables in the HCL => do you mind splitting into PR with different goals to make it easier for othger maintainers to review please (I'm not willing to give and approval with so much changes at the same time).
- Only blocker for now is the missing updated documentation for windows contributor: this PR introduces way more requirements for them and it must be documented.
- I'm not sure the end result is easier: it goes on the opposite style as what we used to do: trying to remove the abstraction of Make/make.ps1 . I won't block for this as it is philosophical
- Could you explain what is the "problem" you want to solve with this PR? As much as it looks interesting and fun technically, I fail to see the gain

File: docker-bake.hcl
Comment (lemeurherve) [reply docker-bake.hcl:94]: The backslashes are part of the string WINDOWS_FLAVORS_TO_BUILD, this isn't an HCL array.

I need them to define the content I want in a JSON format, to be decoded later in an array of flavors to build.

This is the only way I found in order to be able to define array values values via environment variables and to interact with them from the outside.

File: docker-bake.hcl
Comment (lemeurherve) [reply docker-bake.hcl:98]: Same explanation as above.

**Comment by lemeurherve** [comment]
> Could you explain what is the "problem" you want to solve with this PR? As much as it looks interesting and fun technically, I fail to see the gain

The main problem I have with the current status is that we don't have a "pure" definition of Windows images with the docker compose file as it is, we need to massage data with build.ps1 in order to obtain the correct value for TOOLS_WINDOWS_VERSION for example.

This PR proposes having "pure" images definitions, inside a single common file containing all the helper functions (without having to replicate them in build.ps1) and all the parameters needed to generate the images.

It also pave the way for when docker build bake will be usable to build Windows images.

File: Jenkinsfile
Comment (dduportal) [inline Jenkinsfile:50]: It is considered a good Jenkins pipeline practise to avoid direct groovy interpolation like this. You should wrap the `sh` step within a `withEnv` with a literal string passed as argument

**Comment by lemeurherve** [comment]
Put the PR in draft while I'm extracting unrelated changes in another PR and improve the contributor doc.

File: docker-bake.hcl
Comment (lemeurherve) [reply docker-bake.hcl:94]: I've explained (hopefully more clearly) this with a bit more details here: https://github.com/docker/buildx/discussions/2398#discussioncomment-9778966

Reworked this PR with two targets "nanoserver" and "windowsservercore" instead of the "allwindows", and kept only one WINDOWS_VERSION_OVERRIDE variable to be able to set a unique Windows version when calling docker buildx bake to generate the docker compose file.

**Comment by lemeurherve** [comment]
> It looks like you need a lot of additional tools if you are on Windows to be able to build the image (including GNU `make` and `yq`): am I correct?

I reworked the PR to generate the docker compose file from a Windows agent instead of a Linux one, `make` isn't required on Windows anymore.

As `yq` is already required, the only new requirement is https://github.com/docker/buildx, added to the agents with https://github.com/jenkins-infra/packer-images/pull/1215.

> There are a lot of tiny improvement which looks really useful without any doubts, such as factorizations to variables in the HCL => do you mind splitting into PR with different goals to make it easier for othger maintainers to review please (I'm not willing to give and approval with so much changes at the same time).

Extracted in:
- #419
- #425
- #424
- #421

> Could you explain what is the "problem" you want to solve with this PR? As much as it looks interesting and fun technically, I fail to see the gain

I've detailed everything in #426.

> Only blocker for now is the missing updated documentation for windows contributor: this PR introduces way more requirements for them and it must be documented.

Currently working on it, the rest is ready: https://github.com/jenkinsci/docker-ssh-agent/pull/415/commits/7bcc98f061ebc4965c82ac6a5828920220784993

I'll pass the PR in ready for review when I'll add the documentation.

**Comment by lemeurherve** [comment]
Pull request ready for review, this commit especially: https://github.com/jenkinsci/docker-ssh-agent/commit/ed751ae2946c8898350e3b04985c824a017735b1

The rest are cherry-picks from the other pull requests, amongst those #419 and #424 are required to be merged before merging this one.

**Comment by lemeurherve** [comment]
As I put the docker compose file generation into a function, I realized that moving Pester initialization to its own function would result in a more readable build.ps1 structure, hence https://github.com/jenkinsci/docker-ssh-agent/pull/415/commits/cd96ab7e9b4d5256a33d37fdf4af9e3dc1cc4ac6
(can be extracted to its own PR if deemed preferable)

<details>

https://github.com/jenkinsci/docker-ssh-agent/blob/cd96ab7e9b4d5256a33d37fdf4af9e3dc1cc4ac6/build.ps1#L183-L253

</details>

File: Jenkinsfile
Comment (dduportal) [inline Jenkinsfile:47]:
```suggestion
                                    sh 'make buildx-init'
```

nit: the intent is to ensure buildx is initialized, so proposal to (slightly) change the target. The keyword `init` is good and clear 👍

File: Makefile
Comment (dduportal) [inline Makefile:42]:
```suggestion
buildx-init: check-reqs
```

File: Jenkinsfile
Comment (dduportal) [inline Jenkinsfile:61]:
```suggestion
                                            sh 'make build-all'
```

nit: use a dashed case as word separator to be homogeneous with the `ci-init` or `check-reqs` targets

File: Makefile
Comment (dduportal) [inline Makefile:53]:
```suggestion
build-all: check-reqs
```

File: Jenkinsfile
Comment (dduportal) [inline Jenkinsfile:89]: ❤️

File: README.md
Comment (dduportal) [inline README.md:211]:
```suggestion
`buildx-init` is dedicated to Jenkins infrastructure for initializing docker and isn't required in other contexts.
```

Comment (dduportal) [reply README.md:211]: or

```suggestion
`ci-init` is dedicated to Jenkins infrastructure for initializing docker and isn't required in other contexts.
```

File: README.md
Comment (dduportal) [inline README.md:225]:
````suggestion

```console
````

=> In Markdown, it is safer to always surround blocks (code block in this case) by empty lines

**Review by dduportal — APPROVED** [review]
Looks good \o/ 

Devil in the details; I've added a few nitpicks or non blocking comments, I let you decide if you want to keep them or not.

Great job and documentation!

File: docker-bake.hcl
Comment (dduportal) [reply docker-bake.hcl]: That look really clear, thanks!

**Review by dduportal — COMMENTED** [review]
FYI: I'm transplanting the comments in https://github.com/jenkinsci/docker-ssh-agent/pull/421

File: README.md
Comment (lemeurherve) [reply README.md:211]: I chose the English spelling: https://sapling.ai/usage/initialize-vs-initialise, is that OK?

File: README.md
Comment (lemeurherve) [reply README.md:225]: Do you have any reference about hits?
I never encountered any issue on GitHub about putting the code directly under the triple back ticks.

File: README.md
Comment (dduportal) [reply README.md:211]: https://github.com/jenkinsci/docker-ssh-agent/pull/421#pullrequestreview-2124965521

File: README.md
Comment (lemeurherve) [reply README.md:225]: Oh nevermind, I though it was about an empty line below and not above.

**Comment by lemeurherve** [comment]
As the cherry-picked commits and the corresponding (merged) pull requests have diverge, I'm rebasing this pull request into a clean commit.

**Review by dduportal — APPROVED** [review]

**Pull request merged**
